src/cluster/etcd/watchmanager/manager.go

```diff
@@ -127,50 +131,63 @@ func (w *manager) Watch(key string) {
 			}
 		}
 
-		select {
-		case r, ok := <-watchChan:
-			if !ok {
-				// the watch chan is closed, set it to nil so it will be recreated
-				// this is unlikely to happen but just to be defensive
-				cancelFn()
-				watchChan = nil
-				logger.Warn("etcd watch channel closed on key, recreating a watch channel")
+		processWatchUpdate := func(r clientv3.WatchResponse) bool {
+			// handle the update
+			if err = r.Err(); err == nil {
+				w.m.etcdWatchUpdateRcvd.Inc(1)
+				if r.IsProgressNotify() {
+					// Do not call updateFn on ProgressNotify as it happens periodically with no update events
+					return true
+				}
 
-				// avoid recreating watch channel too frequently
-				time.Sleep(w.opts.WatchChanResetInterval())
-				w.m.etcdWatchReset.Inc(1)
+				if err = w.updateFn(key, r.Events); err != nil {
+					logger.Error("received notification for key, but failed to get value", zap.Error(err))
+				}
 
-				continue
+				return true
 			}
 
-			// handle the update
-			if err = r.Err(); err != nil {
-				logger.Error(
-					"received error on watch channel",
-					zap.Uint64("etcd_cluster_id", r.Header.ClusterId),
-					zap.Uint64("etcd_member_id", r.Header.MemberId),
-					zap.Bool("etcd_watch_is_canceled", r.Canceled),
-					zap.Error(err),
-				)
-				w.m.etcdWatchError.Inc(1)
-				// do not stop here, even though the update contains an error
-				// we still take this chance to attempt a Get() for the latest value
-
-				// If the current revision has been compacted, set watchChan to
-				// nil so the watch is recreated with a valid start revision
-				if err == rpctypes.ErrCompacted {
```

Comment:
`ErrNoLeader` or any other non-compaction cancellation will not force a retry, unless `watchChan` gets closed - but it may not happen (see comment below on line 164).